From 243c6d9787ec701b7da74411f58676d301f28277 Mon Sep 17 00:00:00 2001
From: Bogdan-Gabriel Roman <bogdan-gabriel.roman@nxp.com>
Date: Mon, 9 Aug 2021 10:41:26 +0300
Subject: [PATCH 266/269] s32g_storage: read an store boot_source

Cache boot_source variable at get_boot_source() function level such that
any subsequent calls to this function after s32g_io_setup() will not
perform any reads again.

Issue: ALB-7478
Upstream-Status: Pending 

Signed-off-by: Bogdan-Gabriel Roman <bogdan-gabriel.roman@nxp.com>
Signed-off-by: Zhantao Tang <zhantao.tang@windriver.com>
---
 plat/nxp/s32g/s32g_storage.c | 16 +++++++++++-----
 1 file changed, 11 insertions(+), 5 deletions(-)

diff --git a/plat/nxp/s32g/s32g_storage.c b/plat/nxp/s32g/s32g_storage.c
index 8ef43b86c..e63226662 100644
--- a/plat/nxp/s32g/s32g_storage.c
+++ b/plat/nxp/s32g/s32g_storage.c
@@ -141,10 +141,16 @@ static int s32g_check_memmap_dev(const uintptr_t spec)
 	return 0;
 }
 
-static uint32_t get_boot_source(void)
+static uint8_t get_boot_source(void)
 {
-	uint32_t boot_cfg = mmio_read_32(BOOT_GPR_BASE + BOOT_GPR_BMR1_OFF);
-	uint32_t boot_source = (boot_cfg & BOOT_SOURCE_MASK) >> BOOT_SOURCE_OFF;
+	uint32_t boot_cfg;
+	static uint8_t boot_source = INVALID_BOOT_SOURCE;
+
+	if (boot_source != INVALID_BOOT_SOURCE)
+		return boot_source;
+
+	boot_cfg = mmio_read_32(BOOT_GPR_BASE + BOOT_GPR_BMR1_OFF);
+	boot_source = (boot_cfg & BOOT_SOURCE_MASK) >> BOOT_SOURCE_OFF;
 
 	switch (boot_source) {
 
@@ -168,7 +174,7 @@ static void set_fip_img_source(struct plat_io_policy *policy)
 	 * If the previous check had failed, the boot flow would
 	 * not have reached this point.
 	 */
-	uint32_t boot_source = get_boot_source();
+	uint8_t boot_source = get_boot_source();
 
 	/* We know the real FIP image length only after FIP header
 	 * is read and parsed in bl2_plat_handle_post_image_load.
@@ -227,7 +233,7 @@ int plat_get_image_source(unsigned int image_id, uintptr_t *dev_handle,
 
 void s32g_io_setup(void)
 {
-	uint32_t boot_source;
+	uint8_t boot_source;
 
 	if (register_io_dev_memmap(&s32g_memmap_io_conn))
 		goto err;
-- 
2.17.1

